temporalio/nexus/_link_conversion.py

```diff
@@ -0,0 +1,170 @@
+            f"Invalid Nexus link: {link}. Expected path to match {_LINK_URL_PATH_REGEX.pattern}"
+        )
+        return None
+    try:
+        event_ref = _query_params_to_event_reference(url.query)
+    except ValueError as err:
+        logger.warning(
+            f"Failed to parse event reference from Nexus link URL query parameters: {link} ({err})"
+        )
+        return None
+
+    groups = match.groupdict()
+    return temporalio.api.common.v1.Link.WorkflowEvent(
+        namespace=urllib.parse.unquote(groups["namespace"]),
+        workflow_id=urllib.parse.unquote(groups["workflow_id"]),
+        run_id=urllib.parse.unquote(groups["run_id"]),
+        event_ref=event_ref,
+    )
+
+
+def _event_reference_to_query_params(
+    event_ref: temporalio.api.common.v1.Link.WorkflowEvent.EventReference,
+) -> str:
+    event_type_name = temporalio.api.enums.v1.EventType.Name(event_ref.event_type)
+    if event_type_name.startswith("EVENT_TYPE_"):
+        event_type_name = _event_type_constant_case_to_pascal_case(
+            event_type_name.removeprefix("EVENT_TYPE_")
+        )
+    return urllib.parse.urlencode(
+        {
+            "eventID": event_ref.event_id,
+            "eventType": event_type_name,
+            "referenceType": "EventReference",
+        }
+    )
+
+
+def _query_params_to_event_reference(
+    raw_query_params: str,
+) -> temporalio.api.common.v1.Link.WorkflowEvent.EventReference:
+    """Return an EventReference from the query params or raise ValueError."""
+    query_params = urllib.parse.parse_qs(raw_query_params)
+
+    [reference_type] = query_params.get("referenceType") or [""]
+    if reference_type != "EventReference":
+        raise ValueError(
+            f"Expected Nexus link URL query parameter referenceType to be EventReference but got: {reference_type}"
+        )
+    # event type
+    [raw_event_type_name] = query_params.get(LINK_EVENT_TYPE_PARAM_NAME) or [""]
+    if not raw_event_type_name:
+        raise ValueError(f"query params do not contain event type: {query_params}")
+    if raw_event_type_name.startswith("EVENT_TYPE_"):
+        event_type_name = raw_event_type_name
+    elif re.match("[A-Z][a-z]", raw_event_type_name):
+        event_type_name = "EVENT_TYPE_" + _event_type_pascal_case_to_constant_case(
+            raw_event_type_name
+        )
+    else:
+        raise ValueError(f"Invalid event type name: {raw_event_type_name}")
+
+    # event id
+    event_id = 0
+    [raw_event_id] = query_params.get(LINK_EVENT_ID_PARAM_NAME) or [""]
+    if raw_event_id:
+        try:
+            event_id = int(raw_event_id)
+        except ValueError:
+            raise ValueError(f"Query params contain invalid event id: {raw_event_id}")
+
+    return temporalio.api.common.v1.Link.WorkflowEvent.EventReference(
+        event_type=temporalio.api.enums.v1.EventType.Value(event_type_name),
+        event_id=event_id,
+    )
+
+
+def _event_type_constant_case_to_pascal_case(s: str) -> str:
```

Comment:
This is just a generic helper, nothing to do with event types.

```suggestion
def _constant_case_to_pascal_case(s: str) -> str:
```

Reply:
I decided to qualify their names with the `event_type` prefix to avoid any implication that they are suitable for use as general case conversion utilities. For example, they have undefined behavior currently if the input doesn't match the expectation that it is already has the stated case, and the tests don't cover trailing/leading underscores, etc.